File: dashboard/backend/api/ip_rules.py
from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel, Field
from typing import Optional, List
from services.ip_rule_service import IPRuleService
from services.rbac import require_viewer_or_above, require_admin

router = APIRouter(prefix="/api/ip-rules", tags=["ip-rules"])
service = IPRuleService()


class IPRuleCreate(BaseModel):
    ip: str = Field(..., description="IP address or CIDR notation (e.g. 192.168.1.1 or 10.0.0.0/24)")
    rule_type: str = Field("block", description="'block' or 'allow'")
    reason: str = Field("Manual entry", description="Reason for the rule")
    duration_seconds: Optional[int] = Field(None, description="TTL in seconds. None/0 means permanent")
    source: str = Field("manual", description="Origin of rule ('manual', 'ml_analyst', 'rate_limit')")


class BulkDeleteRequest(BaseModel):
    ips: List[str]


@router.get("/")
async def get_ip_rules(
    rule_type: Optional[str] = Query("all", description="'all', 'block', or 'allow'"),
    search: Optional[str] = Query(None, description="Search term for IP or reason"),
    current_user: dict = Depends(require_viewer_or_above),
):
    try:
        rules = service.list_rules(rule_type=rule_type, search=search)
        stats = service.get_stats()
        return {
            "rules": rules,
            "stats": stats
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/")
async def add_ip_rule(
    payload: IPRuleCreate,
    current_user: dict = Depends(require_admin),
):
    try:
        username = current_user.get("username", "admin")
        result = service.add_rule(
            ip=payload.ip,
            rule_type=payload.rule_type,
            reason=payload.reason,
            added_by=username,
            duration_seconds=payload.duration_seconds,
            source=payload.source
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{ip:path}")
async def delete_ip_rule(
    ip: str,
    current_user: dict = Depends(require_admin),
):
    try:
        deleted = service.delete_rule(ip)
        if deleted:
            return {"status": "success", "message": f"Rule for IP {ip} removed"}
        raise HTTPException(status_code=404, detail="IP rule not found")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/bulk-delete")
async def bulk_delete_rules(
    payload: BulkDeleteRequest,
    current_user: dict = Depends(require_admin),
):
    try:
        count = service.bulk_delete(payload.ips)
        return {"status": "success", "deleted_count": count}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: dashboard/frontend/src/api/ipRules.ts
import { api } from './axios'

export interface IPRule {
  id: number
  ip: string
  rule_type: 'block' | 'allow'
  reason: string
  added_by: string
  created_at: number
  expires_at: number | null
  remaining_seconds: number | null
  hits: number
  source: string
}

export interface IPRulesResponse {
  rules: IPRule[]
  stats: {
    total_blocked: number
    total_allowed: number
    temp_banned: number
    permanent_blocked: number
  }
}

export const ipRulesApi = {
  getRules: async (ruleType: string = 'all', search?: string): Promise<IPRulesResponse> => {
    const params = new URLSearchParams()
    if (ruleType && ruleType !== 'all') params.append('rule_type', ruleType)
    if (search) params.append('search', search)
    const res = await api.get<IPRulesResponse>(`/ip-rules/?${params.toString()}`)
    return res.data
  },

  addRule: async (data: {
    ip: string
    rule_type: 'block' | 'allow'
    reason: string
    duration_seconds?: number | null
    source?: string
  }) => {
    const res = await api.post('/ip-rules/', data)
    return res.data
  },

  deleteRule: async (ip: string) => {
    const res = await api.delete(`/ip-rules/${encodeURIComponent(ip)}`)
    return res.data
  },

  bulkDelete: async (ips: string[]) => {
    const res = await api.post('/ip-rules/bulk-delete', { ips })
    return res.data
  },
}
